auth/env_keychain.go

```diff
@@ -12,27 +12,33 @@ import (
 	"github.com/pkg/errors"
 )
 
-// NewKeychain returns either a EnvKeychain or a authn.DefaultKeychain depending on whether the provided environment variable is set
-func NewKeychain(envVar string) authn.Keychain {
+// ResolveKeychain returns either:
+// * a resolved keychain from the ggcr DefaultKeychain or
+// * a multi-keychain with a resolved keychain from the provided environment variable, and a resolved keychain from the ggcr DefaultKeychain
+// depending on whether the provided environment variable is set
+func ResolveKeychain(envVar string, images ...string) (authn.Keychain, error) {
+	defaultKeychain := resolvedKeychain{Auths: buildEnvMap(authn.DefaultKeychain, images...)}
```

Comment:
Summarizing other conversation - since `buildEnvMap` accepts an interface, we could set a package level variable to designate the "default keychain" that will be passed here, and set it to `authn.DefaultKeychain` in our implementation. Then in the tests, we could set that variable to be a fake keychain that implements `authn.Keychain` (this will avoid having main care about this variable).

To simulate the error that occurs when the `DOCKER_CONFIG` directory is not readable, we can have our fake keychain return an error when `Resolve()` is called ([here](https://github.com/buildpacks/lifecycle/blob/ca47d3defd627d7ebb16e5e3b98bb9cc9b2d7ca4/auth/env_keychain.go#L159)). Then we can remove all of the test setup that tried to modify the `DOCKER_CONFIG` directory. Our tests don't need to care at all about `DOCKER_CONFIG`, only that the "default keychain" returned an error.

Reply:
We actually already have a fake keychain for other tests that implements the interface: https://github.com/buildpacks/lifecycle/blob/ca47d3defd627d7ebb16e5e3b98bb9cc9b2d7ca4/auth/env_keychain_test.go#L195

I feel like a quick way to get these tests going would be to add a `returnsForResolve` field on the struct (though it's a little awkward because `Resolve()` returns two values...). Then the `Resolve()` method could check that field, and if set return whatever was provided.

Reply:
@natalieparellano, I mark this conversation as resolved.
Feel free to reopen it if you think that I missed something.

Reply:
@natalieparellano, marking this conversation as resolved.
Feel free to reopen if needed.